janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ * Modifications Copyright 2017 JanusGraph Authors
```

Comment:
Was this file branched from another file in this (or another) repo, and then modified? Would it be possible to create a single commit which is just the clean copy of the file (renamed, perhaps), and the next commit adds the modifications so it's clear what the changes are, or is that too much work / too complicated?

Reply:
I wasn't really sure how to handle this. It borrows heavily from the [TinkerPop SimpleAuthenticator](https://github.com/apache/tinkerpop/blob/3.2.5/gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/auth/SimpleAuthenticator.java)

Maybe it should just list the class it modified? It's also an abstracts away the `JanusGraphSimpleAuthenticator` and the `SaslAndHMACAuthenticator` which also borrow heavily from that class. Would it make sense to list he TinkerPop version and that it borrows from `org.apache.tinkerpop.gremlin.server.auth.SimpleAuthenticator.java`

Reply:
If `credentialsStore` in `SimpleAuthenticator` was protected instead of private (or equivalently if it was obtained through a protected accessor in `SimpleAuthenticator#authenticate`) would you have just been able to collapse `JanusGraphAbstractAuthenticator` and `JanusGraphSimpleAuthenticator` into one, overriding `setup()`, and then update your other implementations to further extend same?

If so then what about in `JanusGraphAbstractAuthenticator#setup` (or after collapse `JanusGraphSimpleAuthenticator#setup`) use reflection to the set `SimpleAuthenticator.credentialsStore`? Reflection isn't very clean but I'd prefer it here over all the code duplication and license header issues, both of which would go away with this update. You could then reach out to TinkerPop about updates to better support auth extensions and contribute that as a separate PR to them so eventually we could remove the need for reflection on the JanusGraph side.

Reply:
@sjudeng This would only address the JanusGraphSimpleAuthenticator, not the `JanusGraphAbstractAuthenticator`. There's also the `SaslAndHMACAuthenticator`, which provides authentication for both HTTP and WebSockets for the combined channelizer (`WsAndHttpChannelizer`). So I could change the `JanusGraphSimpleAuthenticator` to use the TinkerPop `SimpleAuthenticator` through reflection, but I wouldn't be able to _collapse_ the two classes, so it wouldn't fully alleviate the issue. 

It did get me thinking, though, that I'd seen examples of the `GraphFactory` opening JanusGraph properties files. I tested that locally, and it seems to work. We could just open the graph, create the indices, close the graph instance and clear the reference and let the SimpleAuthenticator take care of everything after `super.setup()` is called. I'll give it whirl and check back in with the results.

As far as the license at the top... it is copying a lot, but it's also implementing an interface from TinkerPop. Thinking about it now, I think it would be fine to just include the normal JanusGraph header license. It's implementing the API interface, it's bound to be extremely similar in the way it works. @sjudeng @mbrukman What do you think about that?